### ISSUE TITLE
The compiler should report an error about references to out-of-scope stack memory generated by `[UnscopedRef]` and defensively copied struct.

### ISSUE THREAD

**Issue by ikorin24** [dotnet/roslyn#64776]
**Version Used**: 

.NET7 RC2
C# 11

**Steps to Reproduce**:

Execute the following code.

It occurs accessing to undefined memory (the stack memory already out of scope). 
Remember that there is no `unsafe` keyword in this code.

```cs
using System;
using System.Runtime.CompilerServices;
using System.Diagnostics.CodeAnalysis;
using System.Runtime.InteropServices;

namespace Net7Test;

internal class Program
{
    private static readonly Vec4 ReadOnlyVec = new Vec4(1, 2, 3, 4);

    static void Main(string[] args)
    {
        Console.WriteLine(RuntimeInformation.FrameworkDescription);

        // This refers to stack memory that has already been left out.
        ref Vec4 foo = ref Foo();

        Console.WriteLine(foo);
        MakeStackDirty();
        Console.WriteLine(foo);

        Console.ReadKey();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static ref Vec4 Foo()
    {
        // Defensive copy occurs and it is placed in stack memory implicitly.
        // The method returns a reference to the copy, which happens invalid memory access.
        ref Vec4 xyzw = ref ReadOnlyVec.Self;
        return ref xyzw;
    }

    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.NoOptimization)]
    private static void MakeStackDirty()
    {
        Span<byte> mem = stackalloc byte[1024];
        mem.Clear();
    }
}

public struct Vec4
{
    public float X, Y, Z, W;
    public Vec4(float x, float y, float z, float w) => (X, Y, Z, W) = (x, y, z, w);

    [UnscopedRef]
    public ref Vec4 Self => ref this;

    public override string ToString() => $"<{X}, {Y}, {Z}, {W}>";
}
```

**Expected Behavior**:

The compiler should report an error about accessing to the out-of-scope stack memory
at the method `Foo()`.

**Actual Behavior**:

No compiling error.
We can access the stack memory which is already out of scope, and the program does undefined behaviour.

![bug](https://user-images.githubusercontent.com/19363522/196242938-a8a45aef-4376-4d04-9bb0-2035731dd636.png)

**Comment by ufcpp** [comment]
The ref escape analysis should consider defensive copies of structs.

I found that `in` parameters has the same problem:

```cs
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;

var x = new Wrap { X = 1 };

ref var r = ref m(x);
MakeStackDirty();
Console.WriteLine(r.X); // undefined value

static ref Wrap m(in Wrap i)
{
    ref Wrap r = ref i.Self; // defensive copy
    return ref r; // ref to the local copy
}

[MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.NoOptimization)]
static void MakeStackDirty()
{
    Span<byte> mem = stackalloc byte[1024];
    mem.Clear();
}

struct Wrap
{
    public float X;

    [UnscopedRef]
    public ref Wrap Self => ref this;
}
```

**Comment by RikkiGibson** [comment]
Might want to solve this by adjusting invocation analysis spec a la the following:

If an argument of struct type is a readonly reference and its corresponding parameter is a writable reference, then the escape scope of the invocation is the current local scope (current method or narrower).

We probably need similar language when an `in` parameter is used and an expression which is not a variable is used an argument, i.e. `void M(in Struct s);` M(new Struct())`. This also causes a synthesized local to be created and used in the invocation.

**Comment by jaredpar** [comment]
> We probably need similar language when an in parameter is used and an expression which is not a variable is used an argument, i.e. void M(in Struct s); M(new Struct()). This also causes a synthesized local to be created and used in the invocation.

That language is already there. 

> When discussing the *ref-safe-to-escape* of arguments that correspond to `in` parameters they will be generalized as `ref` arguments in the spec. In the case the argument is an lvalue then the *ref-safe-to-escape* is that of the lvalue, otherwise it is *current method*. Again `in` will only be called out here when it is important to the semantic of the current rule.

This is one case where we use *current method* but mean *current scope*. That is true in a number of places in the spec though because it generalizes the two concepts. 

I think we should add something similar for the receiver of methods.

**Comment by AlekseyTs** [comment]
@RikkiGibson, @jaredpar I am not sure if it is worth chasing this down at the moment. My understanding is that we are planning to rewrite ref-safety analysis as a separate pass. I think that should be done after lowering, when all temps are already present in the bound tree, which, I think, is going to catch the scenario.

**Comment by RikkiGibson** [comment]
It seems like the escape analysis diagnostics would be missed in the IDE if they were moved to occur after lowering. That would be problematic for the user experience.

**Comment by AlekseyTs** [comment]
> It seems like the escape analysis diagnostics would be missed in the IDE if they were moved to occur after lowering. 

I am not sure if this is the case. Even if it is, we are in control what passes are done and when.

**Comment by AlekseyTs** [comment]
There is a question, of course, whether lowering introduces temps for all interesting cases, or IL gen does that instead.
